### DISCUSSION TITLE
Seek command that returns zone names

### DISCUSSION THREAD

**Discussion by RandSim** [ansys/pyfluent#1503] [Q&A] [answered]
Hi,
I have been trying to construct a 'smart' meshing process which would query label/zone names and issue surface refinement depending on said names. Is there a command whose output would be all the zone names in a table or something?
Same question at but for the solver instead of the mesher.
Thanks you for your time!
Alex,

_Skeleton code to help with understanding:_
For **name** in zone_table:
       if **name** contains "_ref03":
            "meshing.workflow.TaskObject['Add Local Sizing'].Arguments.set_state({r'AddChild': r'yes',r'BOIFaceLabelList': [**name**],r'BOISize': <surfRef>,})"

**Comment by RandSim** [comment]
Upping this message so that it doesn't fall through the cracks.

**Comment by mkundu1** [comment]
@RandSim This is ongoing work - please see #1492. We'll add more queries to that PR.

**Reply by RandSim** [reply]
Understood and thank you for the feedback. Is this something I can expect will be available with 2023R2?

**Reply by mkundu1** [reply]
The feature won't be fully available in 2023R2. Meanwhile, we can execute the Fluent meshing queries from PyFluent to get the same info:
```
>>> meshing.scheme_eval.scheme_eval('(tgapi-util-convert-zone-ids-to-name-strings (get-face-zones-of-filter "*let*"))') 
['inlet1', 'inlet2', 'outlet']
```

**Reply by RandSim** [reply]
This isn't working for me, see console below. I directly typed the command provided as is right after loading the CAD in the meshing in Watertight Workflow. The CAD contains labels 'inlet' and 'outlet'. Could you provide more instruction on how to make it work?
Thanks!

"---------------- Import of ball_valve, consisting of 2 parts/objects, complete. 

>>> meshing.scheme_eval.scheme_eval('(tgapi-util-convert-zone-ids-to-name-strings (get-face-zones-of-filter "*let*"))')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'scheme_eval'
>>> "

If `meshing` is your session object returned by `launch_fluent` it should work. Please check what `is None`.

**Reply by RandSim** [reply]
Thanks for all the help!
Here is the output from the fluent meshing console. As far as I can tell, meshing.scheme_eva exists. meshing.scheme_eval.scheme_eval doesn't. 
It would help me a lot if we had an auto complete command feature. Is there such a thing? Tab doesn't autocomplete for me.

>>> meshing
<PyStateEngine.ModelState object at 0x0000000023710A80>
>>> meshing.scheme_eval
>>> meshing.scheme_eval.scheme_eval
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'scheme_eval'
>>>

**Reply by seanpearsonuk** [reply, marked as answer]
As pointed out in my previous reply `meshing` here has to be the object returned by `launch_fluent`. The same instructions do not work otherwise.

**Comment by RandSim** [comment]
Sean,
Thanks for your assistance thus far. Unfortunately, I still have the same behavior as before when I launch the code below with the command launch_fluent as you requested. Any help on this matter would be much appreciated. Perhaps I am missing an import? Perhaps something else? 
Please note that I updated the latest core module for pyFluent with "pip install ansys-fluent-core" and rebooted my computer just to make sure.
Thanks again,
Alex,

The code that I ran is:
"import ansys.fluent.core as pyfluent
meshing = pyfluent.launch_fluent(precision="double", processor_count=8, mode="meshing", show_gui=False)
meshing.workflow.InitializeWorkflow(WorkflowType=r'Watertight Geometry')
meshing.workflow.TaskObject['Import Geometry'].Arguments.set_state({r'FileName': r'C:/Users/alefebvre/Desktop/ClientName/code1/oml157-marketing-Rand.scdoc',"LengthUnit": "mm",})
meshing.workflow.TaskObject['Import Geometry'].Execute()
meshing.scheme_eval.scheme_eval('(tgapi-util-convert-zone-ids-to-name-strings (get-face-zones-of-filter "let"))')"

The output from the code is:
"Traceback (most recent call last):
  File "C:\Users\alefebvre\Desktop\ClientName\code1\Test.py", line 6, in <module>
    meshing.scheme_eval.scheme_eval('(tgapi-util-convert-zone-ids-to-name-strings (get-face-zones-of-filter "let"))')
  File "C:\Users\alefebvre\AppData\Local\Programs\Python\Python310\lib\site-packages\ansys\fluent\core\services\scheme_eval.py", line 312, in scheme_eval
    return self.eval(val)
  File "C:\Users\alefebvre\AppData\Local\Programs\Python\Python310\lib\site-packages\ansys\fluent\core\services\scheme_eval.py", line 245, in eval
    response = self.service.scheme_eval(request)
  File "C:\Users\alefebvre\AppData\Local\Programs\Python\Python310\lib\site-packages\ansys\fluent\core\services\scheme_eval.py", line 57, in scheme_eval
    return self.__stub.SchemeEval(request, metadata=self.__metadata)
  File "C:\Users\alefebvre\AppData\Local\Programs\Python\Python310\lib\site-packages\grpc\_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "C:\Users\alefebvre\AppData\Local\Programs\Python\Python310\lib\site-packages\grpc\_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Error Occurred!"
	debug_error_string = "UNKNOWN:Error received from peer ipv4:192.168.1.57:53064 {grpc_message:"Error Occurred!", grpc_status:2, created_time:"2023-04-28T21:26:56.082385851+00:00"}"
"

**Comment by dnwillia** [comment]
Looks like you are finding the right scheme_eval call but what you are passing it does not look correct compared to what was suggested by mkundu1 earlier:
```
>>> meshing.scheme_eval.scheme_eval('(tgapi-util-convert-zone-ids-to-name-strings (get-face-zones-of-filter "*let*"))') 
['inlet1', 'inlet2', 'outlet']
```
Regarding intellisense, I think at some point this did get broken in VS Code.  What IDE are you using and what version of the library are you using?

**Reply by RandSim** [reply]
Dan, Sean,
Not sure what changed today, but the command below is working fine. Make sure to use the command from a python file (see code below). Do not use it from Fluent command line as it seems that the object "meshing" would be a bit different then.
Hope this helps & thanks everyone for your help,
Alex,


**SOLUTION**
meshing.scheme_eval.scheme_eval('(tgapi-util-convert-zone-ids-to-name-strings (get-face-zones-of-filter "\*let\*"))')


**Code that successfully tested the command**
import ansys.fluent.core as pyfluent
meshing = pyfluent.launch_fluent(precision="double", processor_count=8, mode="meshing", show_gui=False)
meshing.workflow.InitializeWorkflow(WorkflowType=r'Watertight Geometry')
meshing.workflow.TaskObject['Import Geometry'].Arguments.set_state({r'FileName': r'C:/yourPath/yourFile.scdoc',"LengthUnit": "mm",})
meshing.workflow.TaskObject['Import Geometry'].Execute()
test = meshing.scheme_eval.scheme_eval('(tgapi-util-convert-zone-ids-to-name-strings (get-face-zones-of-filter "\*let\*"))')
print(test)
['origin-inlet-nozzle', 'origin-outlet-duct', 'origin-wall-inlet-duct']


**My Environment just for information:**
Python3.10
Pycharm

pip list ansys-fluent-core
Package                               Version
------------------------------------- ---------
ansys-api-fluent                      0.3.5
ansys-api-platform-instancemanagement 1.0.0b3
ansys-fluent-core                     0.12.3
ansys-fluent-parametric               0.6.1
ansys-fluent-visualization            0.6.0
ansys-platform-instancemanagement     1.0.3
appdirs                               1.4.4
certifi                               2022.12.7
charset-normalizer                    3.0.1
contourpy                             1.0.7
cycler                                0.11.0
fonttools                             4.38.0
googleapis-common-protos              1.58.0
grpcio                                1.51.1
h5py                                  3.8.0
idna                                  3.4
imageio                               2.25.1
importlib-metadata                    6.0.0
kiwisolver                            1.4.4
matplotlib                            3.7.0
numpy                                 1.24.2
packaging                             23.0
pandas                                1.5.3
Pillow                                9.4.0
pip                                   23.1.2
pooch                                 1.6.0
protobuf                              3.20.3
protoc-gen-swagger                    0.1.0
pyparsing                             3.0.9
PySide6                               6.4.2
PySide6-Addons                        6.4.2
PySide6-Essentials                    6.4.2
python-dateutil                       2.8.2
pytz                                  2022.7.1
pyvista                               0.38.2
pyvistaqt                             0.9.1
QtPy                                  2.3.0
requests                              2.28.2
scooby                                0.7.1
setuptools                            65.5.0
shiboken6                             6.4.2
six                                   1.16.0
urllib3                               1.26.14
vtk                                   9.2.6
zipp                                  3.14.0
